### ISSUE TITLE
Instanze get stopped after loxone Miniserver uploading new config

### ISSUE THREAD

**Issue by DBadLuck** [UncleSamSwiss/ioBroker.loxone#298]
I recently face the problem, that the ioBroker Loxone instanze gets stopped after I uploaded a new or changed config file to the miniserver.
In the meanwhile the miniserver is rebooting, the ioBroker instanze trys to reconnect to the miniserver but fails. The reboot could take more than 2 minutes. The instance stops after some tries. After the miniserver is back again online/alive after changed config, the iobroker instanze is still stopped.

After that I have to manually restart the ioBroker instanze.

Could this be changed. May changing the option, that connection loss during reboot not cause an unresolved loop.

**Comment by Lothron84** [comment]
I and my neighbour have the same issue!
It‘s very annoying, not to forget to restart the iobroker instance if you change a small thing in the Loxone configuration.

**Comment by raintonr** [comment]
Is this issue still occurring? With the latest adapter version? My miniserver (gen 1) takes a long time to reboot after config change yet the adapter always reconnects correctly.

**Comment by Lothron84** [comment]
Yes, still the same. I use the V 3.0.0

**Comment by raintonr** [comment]
Could you please send any messages you see in the log from the Loxone adapter.

Currently mine logs trying to connect/failed messages constantly while the Miniserver is offline/rebooting but eventually the connect works and all is good.

**Comment by Lothron84** [comment]
All my Miniserver are Gen2.
I have two installations, both with a Client/Gateway Setup of the Miniservers. 

At the first installations the miniserver is used about 30%. This works well with iobroker. 
The second installation is bigger. Miniserver is used by 65% but boots also fast. But here I got the warning after some saving into the miniserver, that a restart loop was detected and iobroker didnt restart the loxone instance. 
See attchments. 

![io1](https://user-images.githubusercontent.com/89299978/224489499-f500396b-f297-4f42-91d9-220302c26b8d.png)

![io2](https://user-images.githubusercontent.com/89299978/224489504-26c6934d-e728-4501-aa8d-43a424ef1b4d.png)

**Comment by raintonr** [comment]
Sorry, but that doesn't give us much to go on.

What machine/OS is IoB running on? How many states are visible to IoB? Are you connecting to Loxone as a user which can see only limited Loxone information or can that user see everything?

Please find the actually log file which should show a proper stack trace and share that.

**Comment by Lothron84** [comment]
I am usaing a Raspberry PI 4 with 2GB Ram.

Where can I see te number of states only of the loxone instance?
The states of all instances are about 8500.

I use the admin account of loxone. So, the user sees all which is set to visible in loxone. 

Do you mean that log? I think the green marked line is the reaseon. iobroker just stopped the restart of the loxone adapter...
![io3](https://user-images.githubusercontent.com/89299978/224493463-65d5a2e8-1ceb-45d9-b1d4-f8476e537044.png)

Or where can I find more detailed logs?

**Comment by raintonr** [comment]
> I am usaing a Raspberry PI 4 with 2GB Ram.

So can I assume Linux? Please login and look in /opt/iobroker/logs. It would be useful to see the rest (the stuff off the right hand side) of those error lines, but...

"Miniserver out of service!" is the clue. I see that is coming from [lxcommunicator](https://github.com/Loxone/lxcommunicator) so will add something to catch that.

**Comment by raintonr** [comment]
I think this is happening when a state change occurs causing `sendCommand` to be called after Loxone has already closed the connection for a reboot. I will add a check for this.

**Comment by raintonr** [comment]
This should be fixed with #402. Please install your adapter direct from here (Github) and let us know if it's working for you.

**Comment by Lothron84** [comment]
Thanks I will test it!

But how can I install a adapter from github? I know only the way in iobroker websurface.

**Comment by Lothron84** [comment]
ok, I get the adapter installed via github. 

I now get several warnings like this:
stateChange loxone.0.1a04f4c5-01b4-4167-ffff5f598a10e034.value while disconnected, discarding

But if I save into the miniserver to often (3-6 times in a row) the adapter is again stopped and not restarted by iobroker. If the info message from iobroker is accepted the loxone instance can be started manually.

For one save into the miniserver the adapter works fine. But if you change some values in the loxone config and safe this to often, the the adpater crashes.

**Comment by Lothron84** [comment]
some logs of the host.iobroker

![io5](https://user-images.githubusercontent.com/89299978/224553614-a14f2c82-f59e-4bf0-9d00-f7e54f8300f5.png)

**Comment by raintonr** [comment]
Hmmmm... OK, so it's coming from somewhere else :disappointed:

We really need to see the end of that 'Caught by controller...' line. Please look in the /opt/iobroker/logs directory on your server and copy/paste the full message from there as text rather than a screenshot.

**Comment by Lothron84** [comment]
ok, i tried to get the logfile. 

But I cant reproduce the exact error. It seam like the command "iob debug loxone.0" didnt restart when the instance crashes. Bit iobroker did that and stops the restart after some tries. 

But I get some exceptions. then the debug output stops. Maybe thats the problem. If the instance wont crash at a save in the miniserver, there should be no problem.

[loxone.log](https://github.com/UncleSamSwiss/ioBroker.loxone/files/10951860/loxone.log)

**Comment by raintonr** [comment]
This is more useful, thanks. Will give it more thought.

**Comment by raintonr** [comment]
Yes, very interesting. These messages show that the Miniserver went offline, 5s later a reconnect was attempted and failed, 5s after that reconnect was again attempted and failed but... the structure file loaded which should never happen!

Moreover, as the structure file did load IoB assumes everything is good and commences sending updates, which fail as the connection is not fully established.

```
Miniserver out of service!
2023-03-12 20:35:25.454  - info: loxone.0 (4110) Socket closed 2008
...
2023-03-12 20:35:30.455  - info: loxone.0 (4110) Trying to connect
2023-03-12 20:35:30.470  - error: loxone.0 (4110) Couldn't open socket
...
2023-03-12 20:35:35.470  - info: loxone.0 (4110) Trying to connect
2023-03-12 20:35:35.486  - error: loxone.0 (4110) Couldn't open socket
2023-03-12 20:35:35.876  - debug: loxone.0 (4110) structure file successfully loaded
2023-03-12 20:35:35.879  - error: loxone.0 (4110) Couldn't enable status updates
2023-03-12 20:35:38.890  - debug: loxone.0 (4110) stateChange loxone.0.1a04f4b6-023c-f1e5-ffff5f598a10e034.value {"val":111.91,"ack":false,"ts":1678649738875,"q":0,"c":"script.js.Shellys.Leistungsmessung_Updates","from":"system.adapter.javascript.0","user":"system.user.admin","lc":1678649738875}
2023-03-12 20:35:38.891  - debug: loxone.0 (4110) Sending command 1a04f4b6-023c-f1e5-ffff5f598a10e034 111.91
2023-03-12 20:35:38.892  - error: loxone.0 (4110) Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2023-03-12 20:35:38.893  - error: loxone.0 (4110) unhandled promise rejection: undefined
2023-03-12 20:35:38.893  - error: loxone.0 (4110) undefined
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
2000
2023-03-12 20:35:38.992  - info: loxone.0 (4110) terminating
```

I will make another branch with the order of the connection logic tweaked for you to test.

**Comment by Lothron84** [comment]
Thanks for your help!

You should also see on the log, that I saved in the miniserver serveral times, but not always the instances crashes. There must be two different behaviors.

**Comment by raintonr** [comment]
No worries... can you please try this on your IoB host (this install make take a minute or two on an RPi):

```
cd /opt/iobroker
npm i raintonr/ioBroker.loxone#issue/298/reconnectChanges
iob stop loxone.0
iob debug loxone.0
```

And let's see what messages come out with that.

**Comment by raintonr** [comment]
@Lothron84 any luck with the new branch?

**Comment by Lothron84** [comment]
Sorry, I was a bit busy the last 2 days.

I tried the new version and saved in the miniserver for about 10 times. The debug output didn't crash like woth the last version! Think this is a good improvment! 
I also dont get a message from iobroker, that that instance will not be restarted again! Also good! 
[loxone2.log](https://github.com/UncleSamSwiss/ioBroker.loxone/files/10983048/loxone2.log)

I attached the debug output! 

I will test more without the debug output activated!

**Comment by raintonr** [comment]
No worries, and that's great news :smile:

Looking at the debug log I noticed one small mistake - forgot to set the `connectionInProgress` flag. There is a double check to prevent multiple connections, but still, that isn't quite right.

Could you please re-install (just `npm i raintonr/ioBroker.loxone#issue/298/reconnectChanges` again) and check that version. You should actually notice no change and it should all still work.

**Comment by Lothron84** [comment]
I just reinstalled the adapter. 
I works for about 8-10 saves into the miniserver. Then it crashes, but this seams to be another problem.
I attached the log.
[loxone3.log](https://github.com/UncleSamSwiss/ioBroker.loxone/files/11011503/loxone3.log)

**Comment by raintonr** [comment]
That last error is a duplicate of #369. Waiting for the upstream merge.

**Comment by Lothron84** [comment]
ok, thanks for your help.

Because my neighbour has the same issue: Do you know, when this changes will be in an offical update?

**Comment by raintonr** [comment]
Fixed in 3.0.1

@DBadLuck @Lothron84 If you are keen to check this you can either install the package direct from NPM or tell IoB to use the 'latest' repository. In fact, I would appreciate if you could test and confirm this release before we mark it as stable.

**Comment by Lothron84** [comment]
Looks very good so far! Thank you!

tested with Loxone config Version 14.
